File: MIDTERM-LAB-3/Java/MP05_ExtractColumn.java
import java.io.*;
import java.util.*;

/**
 * MP05 - Extract Values from a Column
 * Student: PARAON, JOSH HEIDRIC C.
 * Description: Prompts the user for a CSV file path and a column index,
 *              then extracts and displays all values from that column.
 */
public class MP05_ExtractColumn {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Ask user for the CSV file path
        System.out.print("Enter the CSV dataset file path: ");
        String filePath = scanner.nextLine().trim();

        // Ask user which column index to extract (0-based)
        System.out.print("Enter the column index to extract (0-based): ");
        int columnIndex = Integer.parseInt(scanner.nextLine().trim());

        List<String> extractedValues = new ArrayList<>(); // stores extracted column values
        String headerName = "";                           // stores the column header name

        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {

            String line;
            boolean isHeader = true; // flag to identify the header row

            while ((line = reader.readLine()) != null) {
                // Split row by comma, accounting for quoted fields
                String[] fields = parseCSVLine(line);

                if (fields.length <= columnIndex) {
                    // Skip rows that don't have enough columns
                    continue;
                }

                if (isHeader) {
                    // Save the column header name
                    headerName = fields[columnIndex].trim();
                    isHeader = false;
                } else {
                    // Extract the value from the specified column
                    String value = fields[columnIndex].trim();
                    extractedValues.add(value);
                }
            }

            // Display the extracted values
            System.out.println("\n========================================");
            System.out.println("Column Extracted: " + headerName + " (Index " + columnIndex + ")");
            System.out.println("Total Values Extracted: " + extractedValues.size());
            System.out.println("========================================");

            for (int i = 0; i < extractedValues.size(); i++) {
                // Print row number and value
                System.out.printf("[%d] %s%n", i + 1, extractedValues.get(i));
            }

        } catch (FileNotFoundException e) {
            System.out.println("Error: File not found - " + filePath);
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
        } catch (NumberFormatException e) {
            System.out.println("Error: Invalid column index entered.");
        }

        scanner.close();
    }

    /**
     * Parses a single CSV line, handling quoted fields that may contain commas.
     * @param line - the raw CSV line string
     * @return an array of field values
     */
    static String[] parseCSVLine(String line) {
        List<String> tokens = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false; // track whether we're inside a quoted field

        for (char c : line.toCharArray()) {
            if (c == '"') {
                inQuotes = !inQuotes; // toggle quote state
            } else if (c == ',' && !inQuotes) {
                tokens.add(current.toString()); // end of a field
                current.setLength(0);           // reset buffer
            } else {
                current.append(c); // append character to current field
            }
        }

        tokens.add(current.toString()); // add the last field
        return tokens.toArray(new String[0]);
    }
}

File: MIDTERM-LAB-3/JS/MP05_ExtractColumn.js
/**
 * MP05 - Extract Values from a Column
 * Student: PARAON, JOSH HEIDRIC C.
 * Description: Prompts the user for a CSV file path and a column index,
 *              then extracts and displays all values from that column.
 * Run with: node MP05_ExtractColumn.js
 */

const fs = require('fs');
const readline = require('readline');

// Set up readline interface for user input
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

/**
 * Parses a single CSV line, handling quoted fields containing commas.
 * @param {string} line - raw CSV line
 * @returns {string[]} array of field values
 */
function parseCSVLine(line) {
    const tokens = [];
    let current = '';
    let inQuotes = false; // tracks whether we're inside a quoted field

    for (const char of line) {
        if (char === '"') {
            inQuotes = !inQuotes; // toggle quote state
        } else if (char === ',' && !inQuotes) {
            tokens.push(current); // end of a field
            current = '';
        } else {
            current += char; // build current field value
        }
    }

    tokens.push(current); // push the last field
    return tokens;
}

// Step 1: Ask user for CSV file path
rl.question('Enter the CSV dataset file path: ', (filePath) => {
    filePath = filePath.trim();

    // Step 2: Ask user which column index to extract (0-based)
    rl.question('Enter the column index to extract (0-based): ', (indexInput) => {
        const columnIndex = parseInt(indexInput.trim());

        if (isNaN(columnIndex)) {
            console.log('Error: Invalid column index entered.');
            rl.close();
            return;
        }

        // Step 3: Read the CSV file using Node.js File System module
        try {
            const fileContent = fs.readFileSync(filePath, 'utf8');

            // Step 4: Parse CSV into an array of lines
            const lines = fileContent.split('\n').filter(line => line.trim() !== '');

            if (lines.length === 0) {
                console.log('Error: The file is empty.');
                rl.close();
                return;
            }

            // Step 5: Extract header name and values from the specified column
            const headerFields = parseCSVLine(lines[0]);

            if (columnIndex >= headerFields.length) {
                console.log(`Error: Column index ${columnIndex} does not exist. Dataset has ${headerFields.length} column(s).`);
                rl.close();
                return;
            }

            const headerName = headerFields[columnIndex].trim(); // get column header
            const extractedValues = [];                           // stores extracted values

            for (let i = 1; i < lines.length; i++) {
                const fields = parseCSVLine(lines[i]);

                if (fields.length > columnIndex) {
                    extractedValues.push(fields[columnIndex].trim()); // collect column value
                }
            }

            // Step 6: Display formatted output
            console.log('\n========================================');
            console.log(`Column Extracted: ${headerName} (Index ${columnIndex})`);
            console.log(`Total Values Extracted: ${extractedValues.length}`);
            console.log('========================================');

            extractedValues.forEach((value, index) => {
                console.log(`[${index + 1}] ${value}`);
            });

        } catch (err) {
            // Step 7: Handle errors
            if (err.code === 'ENOENT') {
                console.log('Error: File not found - ' + filePath);
            } else {
                console.log('Error reading file: ' + err.message);
            }
        }

        rl.close();
    });
});
